### ISSUE TITLE
ML Transaction Classification Feature

### ISSUE THREAD

**Issue by terrycampbell** [terrycampbell/mybeans#9]
# ML Transaction Classification for MyBeans

## Overview
Add machine learning-based transaction classification to complement the existing rule-based system. Hard-coded rules would be applied first, then ML would suggest classification for unbalanced transactions.

## Current State Analysis

MyBeans already has a sophisticated rule-based classification system with:
- **Rule-based classifiers** with ClassificationRule and ClassificationRuleset
- **Hard-coded account mappings** for specific scenarios (SharedAmexClassifier)
- **Rich training data** in `book/beans/posted_beans/` with 2+ years of transactions
- **Well-structured workflow** (identify → extract → classify → archive)

### 1. Feature Engineering Approach
```python
# Transaction features for ML
features = {
    'text_features': ['narration', 'payee'], 
    'amount_features': ['amount', 'amount_bucket'],
    'temporal_features': ['day_of_week', 'month', 'day_of_month'],
    'account_features': ['source_account', 'historical_patterns']
}
```

### 2. Model Architecture Options

**Option A: Gradient Boosting (Recommended)**
- **Algorithm**: XGBoost/LightGBM
- **Accuracy**: 99%+ based on research
- **Advantages**: Fast training, handles mixed data types, interpretable
- **Use case**: Primary classifier for most transactions

**Option B: Ensemble Approach**
- **Combination**: Random Forest + Text-based classifier
- **Text processing**: TF-IDF vectorization of transaction descriptions
- **Advantages**: Robust to different transaction types
- **Use case**: Fallback when gradient boosting confidence is low

**Option C: Neural Network (Advanced)**
- **Architecture**: BERT embeddings + dense layers
- **Advantages**: Best accuracy for complex patterns
- **Disadvantages**: Requires more data, slower training
- **Use case**: Future enhancement for complex categorization

### 3. Integration Architecture

```python
class MLClassifier:
    def __init__(self):
        self.rule_engine = RuleEngine()  # Existing system
        self.ml_model = load_trained_model()
        self.confidence_threshold = 0.8
    
    def classify_transaction(self, transaction):
        # 1. Apply hard-coded rules first
        rule_result = self.rule_engine.classify(transaction)
        if rule_result.is_complete():
            return rule_result
        
        # 2. Use ML for incomplete transactions
        ml_prediction = self.ml_model.predict(transaction)
        if ml_prediction.confidence > self.confidence_threshold:
            return ml_prediction
        
        # 3. Return incomplete for manual review
        return IncompleteTransaction(transaction, suggestions=ml_prediction.top_3)
```

### Current Limitations
- **Limited algorithms**: Only SVM, no ensemble methods
- **Basic features**: Primarily text-based, limited temporal/amount features
- **No confidence scoring**: Binary classification without uncertainty
- **Single model**: No hybrid rule + ML approach

### Proposed Improvements
1. **Multi-algorithm support**: Add XGBoost, Random Forest options
2. **Rich feature engineering**: Amount patterns, temporal features, account history
3. **Confidence scoring**: Return prediction confidence for manual review decisions
4. **Hybrid classification**: Rule-first, then ML fallback
5. **Continuous learning**: Update model with new classified transactions

### Phase 1: Basic ML Integration (Medium Complexity - 2-3 weeks)
```python
# Add to existing classifier architecture
class MLEnhancedClassifier(Classifier):
    def __init__(self, ml_model_path: str):
        self.ml_model = joblib.load(ml_model_path)
        self.feature_extractor = TransactionFeatureExtractor()
    
    def classify_incomplete(self, transaction):
        features = self.feature_extractor.extract(transaction)
        prediction = self.ml_model.predict_proba(features)
        return prediction
```

### Phase 2: Training Pipeline (Medium Complexity - 1-2 weeks)
```python
class ModelTrainer:
    def __init__(self):
        self.data_loader = BeancountDataLoader()
        self.feature_extractor = TransactionFeatureExtractor()
    
    def train_model(self, beancount_files):
        # 1. Load historical transactions
        transactions = self.data_loader.load_transactions(beancount_files)
        
        # 2. Extract features
        X = self.feature_extractor.fit_transform(transactions)
        y = self.extract_labels(transactions)
        
        # 3. Train XGBoost model
        model = xgboost.XGBClassifier()
        model.fit(X, y)
        
        return model
```

### Phase 3: Advanced Features (High Complexity - 3-4 weeks)
- **Continuous learning**: Update model with new transactions
- **Multi-label classification**: Predict multiple postings per transaction
- **Confidence-based routing**: Route low-confidence predictions to manual review

### Immediate (Week 1-2)
1. **Data preparation**: Extract training data from existing beancount files
2. **Feature engineering**: Implement basic text + amount + temporal features
3. **Model selection**: Train XGBoost on historical data

### Short-term (Week 3-4)
1. **Integration**: Add ML fallback to existing classifier system
2. **Evaluation**: A/B test against rule-based system
3. **Confidence scoring**: Implement prediction confidence thresholds

### Medium-term (Month 2)
1. **Continuous learning**: Retrain model with new classified transactions
2. **Feature enhancement**: Add account history, pattern recognition
3. **GUI integration**: Add ML predictions to dashboard

## Estimated Development Effort
- **Basic ML integration**: 20-30 hours
- **Training pipeline**: 15-20 hours  
- **Advanced features**: 40-50 hours
- **Testing & validation**: 15-20 hours

**Total: 90-120 hours** (suitable for 2-3 month implementation)

## Key Advantages Over smart_importer
1. **Hybrid approach**: Rules first, ML fallback
2. **Better algorithms**: XGBoost vs. SVM
3. **Richer features**: Amount patterns, temporal data
4. **Confidence scoring**: Uncertainty-aware predictions
5. **MyBeans integration**: Native to existing architecture

## Research References
- **smart_importer**: https://github.com/beancount/smart_importer
- **ML Classification Research**: Industry standards show 80%+ accuracy for text-based classification, 99%+ for gradient boosting approaches
- **Feature Engineering**: TF-IDF vectorization, temporal patterns, amount buckets proven effective

This approach leverages the existing rule-based system while adding intelligent ML capabilities for edge cases, providing a robust and maintainable solution.
